testsuite/features/support/env.rb

```diff
@@ -130,26 +135,80 @@ def capybara_register_driver
 # Define the current feature scope
 Before do |scenario|
   $feature_scope = scenario.location.file.split(%r{(\.feature|/)})[-2]
+  if $timeout_failure_count >= $timeout_threshold
+    warn "Timeout failure threshold reached, stopping test suite."
+    exit(1)
+  end
+  feature_path = scenario.location.file
+  $feature_filename = feature_path.split(%r{(\.feature|/)})[-2]
+  next if get_context('user_created') == true
+
+  # Create own user based on feature filename. Exclude core, reposync, finishing and build_validation features.
+  unless feature_path.match?(/core|reposync|finishing|build_validation/)
+    step %(I create a user with name "#{$feature_filename}" and password "linux")
+    add_context('user_created', true)
+  end
 end
 
-# Embed a screenshot after each failed scenario
+# After hook to handle scenario failures and dump feature code coverage into Redis DB
 After do |scenario|
+  # Handle timeout failure threshold
+  if $timeout_failure_count >= $timeout_threshold
+    warn "Timeout failure threshold reached, stopping test suite."
+    exit(1)
+  end
+
   current_epoch = Time.new.to_i
   log "This scenario took: #{current_epoch - @scenario_start_time} seconds"
+
+  # Check if the scenario failed due to a Net::ReadTimeout
   if scenario.failed?
-    begin
-      if web_session_is_active?
-        handle_screenshot_and_relog(scenario, current_epoch)
-      else
-        warn 'There is no active web session; unable to take a screenshot or relog.'
+    if scenario.exception.is_a?(Net::ReadTimeout)
+      $timeout_failure_count += 1
+      warn "Net::ReadTimeout detected! Consecutive ReadTimeouts: #{$timeout_failure_count}"
+      log_server_response_time
+    else
+      begin
+        if web_session_is_active?
+          handle_screenshot_and_relog(scenario, current_epoch)
+        else
+          warn 'There is no active web session; unable to take a screenshot or relog.'
+        end
+      ensure
+        print_server_logs
       end
-    ensure
-      print_server_logs
     end
   end
```

Comment:
Too many if/else statements embedded, I think we need to re-think it.

Reply:
By doing that, you remove the `web_session_is_active?` from the after step. This check save lot of time if an api call is failing.